### PR TITLE
filter_kernels: Minor corrections for filter windowing and radius.

### PR DESCRIPTION
**Pull request by niklata**
These fixes are not particularly visually perceptible in normal use, but correct what I believe to be some mathematical errors in the calculations for video output resampling filter kernels, particularly when downsampling.

bd9a574 should have no effect when upsampling.  85a4125 affects downsampling and upsampling, but only for filters that use intrinsic blur/sharpen such as EWALanczosSharp.

These patches have been tested and they compile properly and produce correct output for me, both when using EWALanczosSharp and temporal Mitchell filtering.

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by haasn** [comment]
Your first change breaks downscaling. The filter radius needs to be widened to account for the extra samples that have to be incorporated to avoid aliasing.

The second change looks good.

**Comment by niklata** [comment]
Yes, the filter does need to sample from a larger width when downscaling, but that should be achieved already by the transformation of the coordinates -- inv_scale is mapping the destination coordinates onto the source coordinates to account for the larger/wider pixels of the source when downsampling.  The underlying filter function itself should still remain unchanged.

As it is, inv_scale is applied and the filter radius is widened.  Only one of these should be done, not both.

See the sample code in http://www-cs.ccny.cuny.edu/~wolberg/pub/crc04.pdf or the relatively clear (compared to ImageMagick) ImageWorsener source https://github.com/jsummers/imageworsener/blob/master/src/imagew-resize.c .

**Comment by haasn** [comment]
> but that should be achieved already by the transformation of the coordinates

The number of samples only depends on `filter->f.radius`. Since you removed the code that makes this larger when downscaling, downscaling no longer widens the filter with your patch.

> The underlying filter function itself should still remain unchanged.

The filter function should be unchanged in the current code. The multiplication in `mp_init_filter` widened the radius and the division in `sample_filter` adjusted the coordinate space.

**Comment by niklata** [comment]
I think I responded before seeing your edit, and looking at correct-downscaling, you are absolutely correct that when the GLSL shaders are built for pass_sample_polar(), the radius in destintation-space must be expanded so that all of the samples are taken in source-space, since it takes the value from scaler->kernel->f.radius.

So the first commit does cause a regression there, irrespective of whatever other merits it may have, since even if it gives corrected values for the filter sample weights, it will cause undersampling.

I think there is still an issue with correctness here, however.  The LUT_POS macro used throughout GLSL code generation calculates a normalized index into the filter LUT in terms of the destination-space, so what I said in my previous comment does seem to apply here, at least to my current reading of things.

For the separable filter sampler, I think my changes in the first commit are correct, and I think that if the use of scaler->kernel->f.radius in pass_sample_polar() is removed (perhaps by passing through the scale_factor set by correct-downscaling to pass_sample_polar()), these changes would also be a correction.  That said, looking at the separable filter sampler, I don't see where it expands the number of samples taken in the source image.  It appears to build a table of sample taps into the weights array, but the number of taps is taken from scaler->kernel->size... which does appear to vary as a function of f.radius!  So you are right here as well.  Samples would be discarded, even if the LUT values are corrected.

So in summary, the first commit will have to be reworked even if has merits.  I won't do so until/unless I hear your opinion of things, since you have much more experience with the mpv code than I do, and it's entirely possible that I'm missing something.

**Comment by niklata** [comment]
And I agree with your analysis.  The existing code is unusual but does appear to be correct.  The only thing I might suggest is a comment explaining the kernel->f.radius dependency in the GLSL code generation and that the adjustment of the coordinate space in sample_filter is intentional.  My first commit is entirely unnecessary.

BTW, thank you for the great job you've done in mpv.  You do excellent work!

**Comment by haasn** [comment]
I'll agree, the code is weird. It grew over time and has gone through multiple iterations as well. Clarifications and documentation would definitely be beneficial, because I usually get confused every time I touch it.

`LUT_POS` is a red herring, it's just there to correct for half-pixel offset in the sample locations. The LUT is essentially just a memoized version of the filter function, ranging from `f(0.0)` to `f(1.0)` for polar filters, and `f(-1.0)` to `f(1.0)` for non-polar filters. LUT_POS is just constructed to make ‘0.0’ align with the first sample position and ‘1.0’ to align with the last sample position.

As for the actual LUT, it's scaled to the source image - 1 unit is one source pixel. The GLSL shader loops over source pixels, adding an offset of 1.0 each time. It's **not** scaled to the destination pixel size. So when downscaling, a single destination will be multiple source pixels, which is why the coordinate space needs to be bigger.

**Comment by haasn** [comment]
Hmm, something I don't actually remember is whether the filter being twice as big means lanczos3 stretched over twice the area or lanczos6 at the normal size.

**Comment by niklata** [comment]
If you'd like, I could try to write some comments.  After your kind help and more digging, I'm starting to get some bearings.  I really should have grepped for f.radius dependencies deeper to start.

There is one slight thing I am unsure of when I went over things more closely to see if I understand properly.  Against my second patch, I believe this may be a correct change:

```
--- a/video/out/filter_kernels.c
+++ b/video/out/filter_kernels.c
@@ -93,7 +93,7 @@ bool mp_init_filter(struct filter_kernel *filter, const int *sizes,
 }

 // Sample from a blurred and tapered window
-static double sample_window(struct filter_window *kernel, double x)
+static double sample_window(struct filter_window *kernel, double x, double inv_scale)
 {
     if (!kernel->weight)
         return 1.0;
@@ -105,7 +105,7 @@ static double sample_window(struct filter_window *kernel, double x)
     x = kernel->blur > 0.0 ? x / kernel->blur : x;
     x = x <= kernel->taper ? 0.0 : (x - kernel->taper) / (1 - kernel->taper);

-    if (x < kernel->radius)
+    if (x < kernel->radius / inv_scale)
         return kernel->weight(kernel, x);
     return 0.0;
 }
@@ -114,8 +114,8 @@ static double sample_window(struct filter_window *kernel, double x)
 static double sample_filter(struct filter_kernel *filter, double x)
 {
     // The window is always stretched to the entire kernel
-    double w = sample_window(&filter->w, x / filter->f.radius * filter->w.radius);
-    double k = sample_window(&filter->f, x / filter->inv_scale);
+    double w = sample_window(&filter->w, x / filter->f.radius * filter->w.radius, 1.0);
+    double k = sample_window(&filter->f, x / filter->inv_scale, filter->inv_scale);
     return filter->clamp ? fmax(0.0, fmin(1.0, w * k)) : w * k;
 }
```

If this looks correct to you, then there may also be a similar issue in

```
static double triangle(params *p, double x)
{
    return fmax(0.0, 1.0 - fabs(x / p->radius));
}
```

Because of p->radius.  I'm not sure anyone who cares about quality would care about a triangle filter for downscaling, though, and it seems troublesome fix.

**Comment by haasn** [comment]
I think the intent was that all of the sampling / window functions return their values at an absolute reference scale.

So for example, the `jinc` window function has its zeroes where you'd expect them, nowhere else. This is why `triangle` also assumes it's going to have its 0 at `x=1` and nowhere else; because the scaling (if any) should be done elsewhere.

**Comment by haasn** [comment]
For example, suppose you have a 3-tap sinc as your kernel and `triangle` as your window, and you're downscaling by a factor of 0.5x.

The original radius of the filter was `3.0`; and inv_scale is 2 (=1/0.5), so f.radius ends up as 6.0.

When sampling, we want ‘x=6’ to end up as our edge, so we need to pass 3.0 to the kernel (x / inv_scale). For the window, we want the edge of the window (radius 1) to align with x=6, which is done by dividing x by `f.radius` (6.0) and multiplying it by `w.radius` (1.0).

**Comment by niklata** [comment]
That sounds correct to me as well.

But for the first snippet, it seems that in the case of the sample_window call for the filter, the zero-cut check compares against filter->radius == base_filter_radius * inv_scale (from mp_init_filter()).  Here we want to cut against the destination-space coordinates, which were properly taken into account for x (as the caller of sample_window() divided it out) but filter->radius still has the implicit inv_scale term multiplied against it (from mp_init_filter()).  The window sample_window() call has no similar issue as there is no inv_scale applied to it so it is still in terms of the base_window_radius.  So at least to me it looks as if that snippet is required to properly enforce a cut-to-zero at the right spot.

**Comment by haasn** [comment]
Yep, your analysis is spot on there. But again, instead of work even more `inv_scale` trickery into the code I believe the proper fix should be to remove inv_scale from `f.radius`. Perhaps it could be split up into “source radius” and “destination radius”, where source radius = destination radius * inv_scale?

The filter kernels would compute off the destination radius, and the GLSL filters / LUT generation would compute off the source radius.

**Comment by niklata** [comment]
I've pushed a new branch https://github.com/niklata/mpv/commits/nk-filter-kernel-fix1 that has the good second changeset from the original pull request and a new changeset that attempts to divide out a dst_radius and src_radius value and remove inv_scale.

Please let me know what you think of it.  It compiles fine and output appears to be good, but these things are tricky to get completely correct and it wouldn't surprise me if I've overlooked something.

I'm a bit unfamiliar with the preferred way for doing things with github pull requests.  Should I close this request and open a new one or is there something else I should do?

**Comment by haasn** [comment]
> I'm a bit unfamiliar with the preferred way for doing things with github pull requests. Should I close this request and open a new one or is there something else I should do?

It depends on the project but the normal way of doing things here would be to force-push over your previous branch.

**Comment by niklata** [comment]
> Edit: When you think about it, the method of sampling we use is equal to taking each destination pixel and placing a copy of the filter function at that position, then multiplying that with the source image. So since pixels are always the same size in the destination space, it should be lanczos3 over twice the area - right?

Yeah, as I understand things, this is precisely correct.

> If so, then I believe we should be able to calculate the filter kernel/window itself regardless of the actual filter size and only scale it in the lut computation, no? It's been a while since I've done any of this, so I'm not really sure.

This is correct too.  The way I think about it is that we're adjusting the coordinates so that the larger pixels of the source (when downsampling) align onto the smaller pixels of the destination.  So as long as we adjust the coordinates, we don't need to alter the filter kernel/window calculations at all.

For the polar/EWA case I'm almost entirely sure things are correct in the current code, at least for the GLSL sampling side.  I got a little confused when I was verifying the LUT generation, but that also looks correct because the mp_compute_weights() function is only called if the sampler function is not polar.  So that looks good, too.

For the 1D separable case, the filter_sizes array in video.c gives us the possible filter->size values.
```
int filter_sizes[] =
    {2, 4, 6, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 0};
```
I think it can technically be marked static even though it isn't now.  And it's passed via reinit_scaler() -> mp_init_filter() and used nowhere else.  So it could be moved to filter_kernel.c.

So in mp_init_filter, we take the next highest integer size of the src_diameter and we check against filter_sizes until we select the first item <= ceil(src_diameter).  So let's consider a inv_scale of 1.5 (downscaling 0.75) with lanczos3.  We'd get ceil(1.5 * 2.0 * 3.0) == 9.0.  Then the next highest selection from filter_sizes would be 12.  The upsampled cases would be ceil(2.0 * 3.0) == 6.0.  I haven't analyzed the GLSL code in detail, but I don't think this filter_sizes array is producing the proper result.  We should be able to simply sample 9 taps.

Double-checking against the polar case, things look fine to me there:

```
void pass_sample_polar(struct gl_shader_cache *sc, struct scaler *scaler)
{
    double radius = scaler->kernel->f.src_radius;
    int bound = (int)ceil(radius);
...
    for (int y = 1-bound; y <= bound; y++) {
        for (int x = 1-bound; x <= bound; x++) {

```

Which is essentially the above without the filter_sizes selection/rounding.

So, perhaps the filter_sizes rounding should be removed?

**Comment by haasn** [comment]
> I think it can technically be marked static even though it isn't now. And it's passed via reinit_scaler() -> mp_init_filter() and used nowhere else. So it could be moved to filter_kernel.c.

Historical baggage. It was technically supposed to be something that the VO can decide; in this case it's tied to OpenGL texture allocation and the way the GLSL code was written. (In the past, there were actually hard-coded paths for the different filter sizes)

The `filter_kernels` stuff is technically supposed to be VO-agnostic.

> We should be able to simply sample 9 taps.

It has to do with the way samples are packed into textures. For sizes of 8 and above, we use RGBA textures where each texel holds four samples, so the LUT size is divisible by 4 (as a function of the number of samples).

But it's possible to rework all of this. We could instead use 1-coordinate (R) textures, like we do for the 1D LUTs. I'm also not sure how `filter->size` and `filter->src_radius` actually interact in the case you described. Would the src_radius be 9 even though the size is 12?

**Comment by niklata** [comment]
I see now; the different behavior for filter sizes in the 1D/separable filter code was confusing me, as I am used to using 1-coordinate (R) textures for that purpose.  Your explanation will make it a lot easier for me to understand that part of the code generation.

My (possibly flawed) understanding is that the current code will generate correct output as it is; it is just sampling more inputs than are strictly necessary when downsampling, which incurs a performance cost but maintains quality.  This understanding is contingent on the number of source samples in the current code being 12 given the example I gave.

Mathematically, the minimum number of samples that should be taken in the case I described is 9.  The current code appears to use a `filter->size` of 12 in that case.

I think the issue with axing src_radius is in the polar/EWA code, which doesn't use `->size` at all (it is always set to 1 in this case).  As it is, the 1D and polar/EWA code are almost totally different code paths.

**Comment by niklata** [comment]
I've pushed a commit that removes `src_radius` and just uses `filter->size` instead.  The meaning differs slightly between EWA and 1D resampling: for EWA, it is the radius in source units, and for 1D it is its original meaning of N source samples.

This required one small change in video/out/opengl/video.c to compensate for `filter->size` not always being equal to `1` in the polar case.  Now it just simply checks `scaler->kernel->polar` directly and sets the enforced value of `1` there.

inv_scale also comes back because it is used in `mp_compute_weights()`, but that is fine since it had to be derived from the radii as well.

If this commit looks good to you, I can squash it into the prior commit (and also rename `dst_radius` to `radius`, which I just remember you suggested but I forgot to do here).  But it's probably easier to review this delta patch first.

**Comment by niklata** [comment]
I pushed a merged commit that should be a lot nicer.  I've tested it against a horizontal-panning scene against a stable build of mpv with temporal interpolation and I don't notice any change in behavior.

This commit does as you (correctly) suggested and removes `src_radius` entirely and simply uses `radius` as the filter radius and calculates what was src_radius at the locations where it is required using `inv_scale`.

It looks a lot nicer to my eyes, too.  Please let me know what you think.

**Comment by haasn** [comment]
> My (possibly flawed) understanding is that the current code will generate correct output as it is; it is just sampling more inputs than are strictly necessary when downsampling, which incurs a performance cost but maintains quality

I think you're right. In the current code, the radius of the actual filter window (as computed with `sample_filter`) was still based on `inv_scale`, and since `inv_scale` doesn't get rounded up the way `filter->size` does it would still end up as 9 nonzero contributions (despite going through 12 samples). It is a performance loss but should not affect the output.

> It looks a lot nicer to my eyes, too. Please let me know what you think.

After giving it a read, I guess neither `src_radius` nor `inv_scale` are really any different from each other, although keeping `inv_scale` avoids redundant changess-for-the-sake-of-change. Maybe we could even rename it to something like `filter_scale` to make it clearer when reading the code inside `pass_sample_polar` and so on why it's actually being multiplied into the filter radius; and why it's being divided out of ‘x’ when generating the LUT. That also makes it clear that `filter_scale` is not really the inverse of the scale factor. I would be happiest with that name change.

As for `size`, it was always intended as a field to contain the width of the actual LUT for the purposes of texture allocation; so keeping it as 1 for EWA definitely makes sense because those LUTs are indeed 1 sample wide.

**Comment by niklata** [comment]
I've renamed `inv_scale` to `filter_scale`.  I also think that `inv_scale` is a confusing name.

I've corrected the bad behavior I noticed before when testing temporal upsampling.  It was a latent-landmine that I've commented.  I also tested using the (ringier, so easier to identify possible regressions) temporal Lanczos, and it looked unchanged from stable mpv to my eyes.

Please let me know if you find any issues.

File: video/out/filter_kernels.c [hunk 1]
```diff
@@ -54,25 +55,26 @@ const struct filter_kernel *mp_find_filter_kernel(const char *name)
 }
 
 // sizes = sorted list of available filter sizes, terminated with size 0
-// inv_scale = source_size / dest_size
+// filter_scale = source_size / dest_size
 bool mp_init_filter(struct filter_kernel *filter, const int *sizes,
-                    double inv_scale)
+                    double filter_scale)
```
Comment (haasn) [inline video/out/filter_kernels.c]: To clarify: I would keep the parameter named “inv_scale”, since it comes from the actual scale factor. “inv_scale” turns into “filter_scale” by clamping it to values above 1.

File: video/out/filter_kernels.h [hunk 2]
```diff
@@ -34,7 +34,7 @@ struct filter_kernel {
     bool polar;         // whether or not the filter uses polar coordinates
     // The following values are set by mp_init_filter() at runtime.
     int size;           // number of coefficients (may depend on radius)
-    double inv_scale;   // scale factor (<1.0 is upscale, >1.0 downscale)
+    double filter_scale;   // scale factor (<1.0 is upscale, >1.0 downscale)
```
Comment (haasn) [inline video/out/filter_kernels.h]: Comment here also sort of doesn't fit the new name; `filter_scale` is more like `how much wider than normal to stretch the filter`.

**Comment by niklata** [comment]
Agh, I pushed before rebasing.  Let me correct that.

**Comment by niklata** [comment]
> As for size, it was always intended as a field to contain the width of the actual LUT for the purposes of texture allocation; so keeping it as 1 for EWA definitely makes sense because those LUTs are indeed 1 sample wide; so I don't necessarily agree with the “filter->polar ? 1 : filter->size” change.

I've removed this change.  It was essentially an artifact of trying to use `filter->size` in the EWA GSL generation instead of `filter->f.radius` and `filter->filter_scale`.

File: video/out/opengl/video.c [hunk 3]
```diff
@@ -3400,6 +3400,9 @@ void gl_video_configure_queue(struct gl_video *p, struct vo *vo)
         const struct filter_kernel *kernel =
             mp_find_filter_kernel(p->opts.scaler[SCALER_TSCALE].kernel.name);
         if (kernel) {
+            // inv_scale wouldn't be correctly initialized were we to use it here.
+            // This is fine since we're always upsampling, but beware if downsampling
+            // is added!
```
Comment (haasn) [inline video/out/opengl/video.c]: Minor note: It's called `filter_scale` now, comment still calls it `inv_scale`

Comment (niklata) [inline video/out/opengl/video.c]: Good spot!  My brain tends to auto-filter comments when reading code so I'm bad about missing these things.  Will fix.

**Comment by haasn** [comment]
Looks good to me now. Thanks for your interest in improving this code and also verifying the current behavior.

**Comment by niklata** [comment]
Thank you for your patience and fantastic explanations of things I didn't understand in the code!

**Comment by atomnuker** [comment]
Looks okk to me as well
Thanks, pushed
